add export of cpl_scalars

DeniseWorthen opened the pull request:
Export 2d-size of tile and number of tiles for use by CMEPS mediator history files. 

See additional information in UFS Issue https://github.com/ufs-community/ufs-weather-model/issues/2171

DeniseWorthen committed:
add export of cpl_scalars

* export 2dsize of tile and number of tiles for use by CMEPS
mediatory history files

DeniseWorthen commented:
@uturuncoglu This is working fine for UFS. I am able to get the land history files on 6 tiles. I have a run here

```
 /glade/derecho/scratch/worthen/FV3_RT/rt_26609/lnd.db
```

The file ``ufs.cpld.cpl.hi.2021-03-22-43200.nc`` shows the fields on the 6 tiles.

uturuncoglu commented:
@DeniseWorthen I am not sure there is an issue or not but please see following plots. The first one is `q2m` and from land history (`ufs.cpld.lnd.out.2021-03-23-21600.tile1.nc`) and the second one is `atmExp_Sl_q` and from mediator history (`ufs.cpld.cpl.hi.2021-03-23-21600.nc`). The mediator one seems lots of zeros over land. 

![Screenshot 2024-03-08 at 10 39 24 AM](https://github.com/NOAA-EMC/noahmp/assets/1228999/9c8be80f-7c51-4131-aa83-feac284d1133)
![Screenshot 2024-03-08 at 10 39 24 AM 2](https://github.com/NOAA-EMC/noahmp/assets/1228999/27a35fb6-192c-43f3-bec2-c0454ab86f00)

I am not sure we have same issue with original implementation or not. It seems other variables have same issue but since the units are different in model history and mediator history, it is hard to tell.

uturuncoglu commented:
The following is for zvfun
![Screenshot 2024-03-08 at 10 47 57 AM](https://github.com/NOAA-EMC/noahmp/assets/1228999/a25117f5-52b0-4d50-baa6-5b4635d057f5)
![Screenshot 2024-03-08 at 10 47 57 AM 2](https://github.com/NOAA-EMC/noahmp/assets/1228999/0ff28545-74fc-41ee-b32e-67a4ef5f38b4)

uturuncoglu commented:
@DeniseWorthen It would be also nice to run `/glade/derecho/scratch/worthen/FV3_RT/rt_26609/control_p8_atmlnd_intel` case again by adding those options for the atm output. So, we could also check the fields exported to land component.

DeniseWorthen commented:
Hm....the ``ufs.cpld.lnd.out.2021-03-23-21600.tile1.nc`` files are also export state variables on tiles? 

One test I didn't do is to check using the current method of setting the tilesize via config. I think you had used that previous. The new history files should be equivalent.

uturuncoglu commented:
@DeniseWorthen Yes. The land output also includes the forcing fields coming from atmosphere. So, for example `dswsfc` is the shortwave, `t1` is temperature and `q1` is humidity. Wind is little bit different since we are not using wind direction.

uturuncoglu commented:
@DeniseWorthen Yes, if I remember correctly, I was using config options also for land but it is hard to remember exactly.

DeniseWorthen commented:
@uturuncoglu I made a couple of checks. 

First, I wrote the mediator history files on the mesh and decomposed them into the tiles using NCL. Those fields match the import/export fields I get if I use the original method of setting a config variable

```
  history_tile_atm = 96
  history_tile_lnd = 96 
```

This original method only worked for component history files, but even when the fields are split up into two files (one for atm history and one for lnd history), the fields are identical to the original mesh fields.

Then I checked that the fields using the new ``cpl_scalars``  and the fields from NCL are the same as those written directly via CMEPS. So I'm pretty confident the new ``cpl_scalars`` are writing the tiled history files correctly. I *did* find an issue that the fractions are not written on the tiles when you use a single history write phase, so I'll need to fix that in CMEPS. I can fix that later though.

Second, I looked at time=25200 on tile 1 and the zvfun in the "out" file compared to the lndImp_Sl_zvfun and the atmexp_Sl_zfun  in the mediator history.

<img width="400" alt="Screenshot 2024-03-08 at 4 55 07 PM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/c14e78c0-f509-457a-8e47-ad6dbc76123d">

<img width="400" alt="Screenshot 2024-03-08 at 4 55 23 PM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/ea52e120-f3e4-4a8e-9fe4-e2f1114d4fa6">

<img width="400" alt="Screenshot 2024-03-08 at 5 10 15 PM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/5ae315de-e260-4c5b-9356-240b9d1e88d4">

So the lndimp looks the same as the lnd out file, but the atmexp is different. I don't know whether that is expected.

I also checked the DSWSFC field from the "out" and the LNDEXP_FAXA_SWDN and ATMIMP_FAXA_SWDN fields and those all look consistent.

uturuncoglu commented:
@DeniseWorthen Thanks for checking. Yes, it seems that there is an issue in atm export. I think that `lndImp_Sl_zvfun` needs to be same as `atmexp_Sl_zfun`. I am not sure why we have this difference. Is it same for the out-of-box configuration? I mean without any modification. We could check the baseline if this was an issue in the first place. If so maybe there is a bug in mediator since `lndImp_Sl_zvfun` is fine but `atmexp_Sl_zfun` not.

uturuncoglu commented:
@DeniseWorthen @barlage I was also looking to the debug mode issue. I put required checks to the place that we have issue and still failing. In think I know why, we are using `C96.vegetation_type.tile` to get vegetation type and using it (along with fraction) to construct the land sea mask but we are also using initial condition `sfc_data.tile` files to get vegetation type and I think that the vegetation type information found in `C96.vegetation_type.tile` is not consistent with the one found in `sfc_data.tile`. Look at attached plot for the difference of vegetation type for tile 1 (static file - src file)

![Screenshot 2024-03-08 at 3 54 17 PM](https://github.com/NOAA-EMC/noahmp/assets/1228999/a88af5b1-d9e9-4556-98fd-f4468785937c)

There are lots of differences and I think there are also land-sea mask difference between those files. And it leads to issue. I could modify the code and make it consistent by using also src file to construct land-sea mask. Just to be sure, I also need to check datm coupled configuration since that uses custom initial condition. @barlage let me know what you think? @DeniseWorthen I think once I fix the issue, I think this could be part of your PR since you have some modification in the land model too. Of course we also need to solve the issue in `atmexp_Sl_zfun` and also other `atmexp_` fields.

uturuncoglu commented:
@DeniseWorthen @barlage I think this is also related to the https://github.com/ufs-community/ufs-weather-model/issues/1423

uturuncoglu commented:
@DeniseWorthen @barlage Okay. I fixed the issue with vegetation fraction but now getting error from soil type. This is also read from `C96.soil_type.tile*.nc` files but also found in `sfc_data.tile*.nc` (I think stype, btw it is hard to know which variable which since surface file has no meaningful long name in it and just variable name. Anyway, maybe I could get the variable from surface file but it won't be consistent with others. Anyway, maybe for active atmosphere coupled configurations the all information needs to be populated from surface file and for data atmosphere configuration I could use static files (start with C96). I am getting following variables from static files,

```
geolat - oro_data.tile*.nc
soil_type - soil_type.tile*.nc
vegetation_type - .vegetation_type.tile*.nc (this also needs to be read from sfc)
slope_type - .slope_type.tile*.nc
substrate_temperature - substrate_temperature.tile*.nc
maximum_snow_albedo - maximum_snow_albedo.tile*.nc
vegetation_greenness - vegetation_greenness.tile*.nc
soil_color - soil_color.tile*.nc
```

but I think surface file does not have all of those variables and we will end up with inconsistencies on those variables in terms of land-sea mask and actual values like I showed for zvfun.

DeniseWorthen commented:
> @DeniseWorthen Thanks for checking. Yes, it seems that there is an issue in atm export. I think that `lndImp_Sl_zvfun` needs to be same as `atmexp_Sl_zfun`. I am not sure why we have this difference. Is it same for the out-of-box configuration? I mean without any modification. We could check the baseline if this was an issue in the first place. If so maybe there is a bug in mediator since `lndImp_Sl_zvfun` is fine but `atmexp_Sl_zfun` not.

I did run the develop branch and use NCL to write out the fields on the tiles. I see the same thing---the lndImp and the atmExp of Sl_zfun are not the same, and I agree they should be. But, I also don't understand what the field "land fraction" fields mean. The lfrac in the atm looks the same as the lfrac in the land, but I don't understand how lfrac is zero on land. It must mean something different than I think:

<img width="400" alt="Screenshot 2024-03-09 at 11 37 45 AM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/ec3d38d8-cb6c-4c76-b009-4d24b6facc1f">

uturuncoglu commented:
@barlage @DeniseWorthen I made some changes in the code to solve the issue under debug mode which is triggered by inconsistencies in the input data. So, following is the initial attempt to solve it,

https://github.com/NOAA-EMC/noahmp/compare/develop...NOAA-EMC:noahmp:hotfix/debug_mode?expand=1

This is working and passing the point that Denise had issue but failing with another error like following, 

```
forrtl: severe (408): fort: (3): Subscript #1 of the array ALBSAT_TABLE has value -999 which is less than the lower bound of 1

Image              PC                Routine            Line        Source
fv3.exe            000000000AB02772  noahmpdrv_mp_tran        1402  noahmpdrv.F90
fv3.exe            000000000AADFE10  noahmpdrv_mp_noah         853  noahmpdrv.F90
fv3.exe            000000000E289048  lnd_comp_driver_m         601  lnd_comp_driver.F90
fv3.exe            000000000E25EF9C  lnd_comp_nuopc_mp         458  lnd_comp_nuopc.F90
```

This is basically caused by the soil color data. If you look at my changes closely, I am still getting soil color data from static file rather than sfc file and this causes issue since those are not consistent. I am not sure how this can be solved. Two possible options are (1) create new initial condition (sfc file) that has also other variables line soil color (I still don't sure what the status soil color does not match with sfc file in terms of masking) or (2) update soil color static file. @barlage let me know what you think?

uturuncoglu commented:
@DeniseWorthen I think this is not related with the land fraction field and I think we have issues with other fields. Maybe some recent changes in the CMEPS side. The baseline created after land PR is,

`/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/develop-20240124/control_p8_atmlnd_intel/`

and current baseline is,

`/glade/derecho/scratch/epicufsrt/ufs-weather-model/RT/NEMSfv3gfs/develop-20240301/control_p8_atmlnd_intel/`

The `control_p8_atmlnd_intel` test does not check the mediator history file. So, it is not easy to check them quickly. I could try to run the version of ufs-weather-model just after land PR `https://github.com/ufs-community/ufs-weather-model/tree/932a532516e8b9fd7aa13cc137f777cbcc8e8709` and check it.

BTW, if I check the FV3 output from your run `/glade/derecho/scratch/worthen/FV3_RT/rt_26609/control_p8_atmlnd_intel/sfcf024.tile1.nc` and look at `spfh2m` it looks strange and this is also same from the `develop-20240124`. So, maybe we missed the issue and CMEPS was buggy also in the land PR. Anyway, I'll look at the issue.

uturuncoglu commented:
@DeniseWorthen Maybe there is something wrong with land frac. So, that leads issue when data is transferred from land to atmosphere. Anyway, I am trying to run in my side to test some ideas.

DeniseWorthen commented:
@uturuncoglu Maybe Mike understands better, but I know George recently committed a new "v2" surface file test. I don't know if those files are more appropriate to use (the old land mask mis-match issue?). 

I feel like I opened a can of worms but I do think the cpl_scalars are working correctly (except for  https://github.com/ESCOMP/CMEPS/pull/436#issuecomment-1986915263). Let me know what I can do to help.

uturuncoglu commented:
@DeniseWorthen Yes, file inconsistencies are the main issue in here. I am glad that we have issue and It force us to fix it. We have weekly tag-up with @barlage today. So, we could discuss it. I agree it seems that I/O part seems working. I am still working on land fraction issue, I might solve it today.

uturuncoglu commented:
@DeniseWorthen I investigate it little bit further. Here is my finding, the following part of the code in CMEPS basically assigns the lfrac, https://github.com/NOAA-EMC/CMEPS/blob/624920ddbd819c76ec37591c24e872308201810e/mediator/med_fraction_mod.F90#L257. I checked both `is_local%wrap%FBImp(complnd,complnd)` and `is_local%wrap%FBFrac(complnd)` and it seems those are fine. But, the one (https://github.com/NOAA-EMC/CMEPS/blob/624920ddbd819c76ec37591c24e872308201810e/mediator/med_phases_post_lnd_mod.F90#L61) used in the `med_map_field_packed` routine is not. Some how the land fraction used in the run phase is corrupted. Anyway, at this point I have no idea why it is corrupted in the post phase. I'll keep checking to find the reason.

DeniseWorthen commented:
@uturuncoglu good digging. I wonder what it means when it says as a comment "this may be overwritten later" (in the fraction_mod).

uturuncoglu commented:
@DeniseWorthen Yes. In the fraction file, the mediator trying to map land fraction coming from atmosphere as land fraction. I put more debug code to cmeps to find out this is the issue or not. If so, we might need to put some specialization in here. I am not sure configuration under CESM. Maybe there is some implicit assumption over there. I think FV3 does not export land fraction but need to check.

barlage commented:
@DeniseWorthen @uturuncoglu I did a little digging into the sfc_data and fix files being used in the RT. What I mean below by "not consistent" is that there are not valid land characteristics over **all** grids where land_frac>0. From what I can tell:

- those used in the RT by default are not consistent, e.g.
```
fixpath = "/scratch1/NCEPDEV/stmp2/Michael.Barlage/FV3_RT/rt_114204/control_p8_atmlnd_intel/INPUT/"
sfcpath = "/scratch1/NCEPDEV/stmp2/Michael.Barlage/FV3_RT/rt_114204/control_p8_atmlnd_intel/INPUT/"
```
- I believe these are coming from here and they are also not consistent (doesn't matter which oro file is used):
```
fixpath = "/scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20221101/FV3_fix_tiled/C96/"
sfcpath = "/scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20221101/FV3_input_data/INPUT_L127/"
```
**Note these are consistent for non-fractional grids.**
- these "v2" data are consistent for fractional grids and probably should be used:
```
fixpath = "/scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20221101/FV3_input_data/INPUT_L127_v2_sfc/fix_sfc/"
sfcpath = "/scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20221101/FV3_input_data/INPUT_L127_v2_sfc/"
```
- these also seem consistent, but may be dangerous to use since they are a mix of two dirs and may not be consistent in other ways
```
fixpath = "/scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20221101/FV3_input_data/INPUT_L127_v2_sfc/fix_sfc/"
sfcpath = "/scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20221101/FV3_input_data/INPUT_L127_mx100_v2_sfc/"
```

uturuncoglu commented:
@barlage Thanks for checking. Since the test uses `tests/fv3_conf/control_run.IN` file to copy files, I check that one and there is a special variable `V2_SFC_FILE`that needs to be set to `.true.` to use V2 files. Anyway, let me try to include that variable to land test and see what happens.

uturuncoglu commented:
@DeniseWorthen Following part of code changes the land fraction. https://github.com/NOAA-EMC/CMEPS/blob/624920ddbd819c76ec37591c24e872308201810e/mediator/med_fraction_mod.F90#L454. It seems that  'lfrac' in FBFrac(compatm) is set by https://github.com/NOAA-EMC/CMEPS/blob/624920ddbd819c76ec37591c24e872308201810e/mediator/med_fraction_mod.F90#L390 like following 

```
          if (associated(lfrac)) then
             if (is_local%wrap%comp_present(complnd)) then
                do n = 1,size(lfrac)
                   lfrac(n) = 1.0_R8 - ofrac(n)
                   if (abs(lfrac(n)) < eps_fraclim) then
                      lfrac(n) = 0.0_R8
                   end if
                end do
             else
                lfrac(:) = 0.0_R8
             end if
```

using ocean fraction but since we don't have ocean component and using `eps_fraclim` it is going weird. I am plaining to put some control in here and slip section start with L454 since we have land fraction coming from land and no ocean component. @mvertens Is there any CESM configurations that just couple atmosphere with land without any ocean component (data or active). Maybe this configuration is never tested under CESM. @DeniseWorthen @mvertens Please let me know what you think?

uturuncoglu commented:
@DeniseWorthen Okay. The part that sets the land fraction using ocean is just used when either ocean and ice model present. I think we need to add extra protection to 822L to prevent getting land fraction from atmospheric model component.

uturuncoglu commented:
Sorry line 461 in `med_fraction_mod.F90`. Anyway, I also checked the land fraction coming from atmospheric model component and it seems it is corrupted. See following plot,

![Screenshot 2024-03-11 at 3 56 56 PM](https://github.com/NOAA-EMC/noahmp/assets/1228999/1e5f9b6f-117a-4811-929a-c08efeb525a0)

As you can see, we have lots of zeros over land which is not realistic and created issue. I think I our case, we have to use land provided fraction rather than the one interpolated from atmosphere which is not correct for UFS case.

uturuncoglu commented:
@DeniseWorthen Maybe the mapping in https://github.com/NOAA-EMC/CMEPS/blob/624920ddbd819c76ec37591c24e872308201810e/mediator/med_fraction_mod.F90#L431 is not working correctly in out case since it is trying to map land frac from land component to atm.

uturuncoglu commented:
@DeniseWorthen Okay. I traced the issue more. It seems that atmospheric model mask shows the same issue. Mediator log has following,

```
 (module_med_map: med_map_routehandles_initfrom_field) creating RH consd for lnd to atm srcMask =    -987987 dstMask =          1
```

So, since destination mask (fv3 model) is corrupted the interpolation has also issue. I checked the fv3 configuration and it seems it sets the mask in `addLsmask2grid` under `FV3/atmos_model.F90` like following,

```
!$omp parallel do default(shared) private(i,j,nb,ix)
    do j=jsc,jec
      do i=isc,iec
        nb = Atm_block%blkno(i,j)
        ix = Atm_block%ixp(i,j)
! use land sea mask: land:1, ocean:0
        lsmask(i,j) = floor(one + epsln - GFS_data(nb)%SfcProp%oceanfrac(ix))
      enddo
    enddo
```

also `oceanfrac` is set under `FV3/io/fv3atm_sfc_io.F90` and differently for fractional grid or not. In my case, `frac_grid    = .true.`. So, I am not sure what is wrong in here. I keep digging. Now, I started to think that maybe setting atmospheric model mask has issue.

DeniseWorthen commented:
@uturuncoglu What I still don't understand is what land fraction should be on land, away from a coast. Near a coast, "land frac" in FV3 has a meaning relating to the mapped ocean mask. That is the 1-land_frac in the below figure (it is called land_frac, but it is really ocean_frac) used in the setting of the lsmask variable. 

<img width="400" alt="Screenshot 2024-03-11 at 7 03 05 PM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/d59c61d7-f101-4822-a370-1aa8f59c5189">

Does the land model have a concept of "fractional land", so that a grid cell in the middle of the Sahara would be something other than land_frac=1? Your satellite view figure above agrees w/  the variable ``med_frac_atm_lfrac`` I posted above. It seems like the imported field (lndImp_Sl_zvfun) is getting masked by this atm_lfrac before being sent to the atm. I just don't know what the land frac it should be using is.

DeniseWorthen commented:
> @DeniseWorthen Maybe the mapping in https://github.com/NOAA-EMC/CMEPS/blob/624920ddbd819c76ec37591c24e872308201810e/mediator/med_fraction_mod.F90#L431 is not working correctly in out case since it is trying to map land frac from land component to atm.

This test in question (atmlnd_p8) is running LND on the same grid as ATM, right ? So shouldn't it just be doing a redist (the mapcopy)?

uturuncoglu commented:
@DeniseWorthen I checked the CESM B compost which has all active components. Here are the values of the variables

Med_frac_lnd_lfrac - 1 over land 0 over ocean
Med_frac_atm_lfrac - same with Med_frac_lnd_lfrac

I think this is also consistent with your plot. There are some points round lakes has land fraction less then one but not like the plot that I create previously. That has definitely issue and I think it is not related with the fraction since I think model sending it correctly. The issue appears when CMEPS perform interpolation to transfer the fraction from land grid to atmosphere grid (I am not sure why it is not copy but in general the interpolation must run since you could run the configuration with higher resolution land like C384). In this case, it seems that atmosphere mask has issue and that leads to an issue in the interpolation and some regions over land are masked out.

uturuncoglu commented:
@DeniseWorthen Okay. I write out component meshes in CMEPS side using ESMF_MeshWrite call. It seems that mask coming from FV3 has issue. Following is the mask value from FV3 mesh.

![Screenshot 2024-03-11 at 10 09 27 PM](https://github.com/NOAA-EMC/noahmp/assets/1228999/352bd2a6-1907-4df5-bc83-c0f4c2883eac)

I think I need to go to https://github.com/NOAA-EMC/fv3atm/blob/062f7958ccd9bd97b8737432f75b06ec734777a7/io/fv3atm_sfc_io.F90#L1101 and try to understand logic over there. There might be an issue in Sfcprop(nb)%oceanfrac definition since it is used to set masking in https://github.com/NOAA-EMC/fv3atm/blob/062f7958ccd9bd97b8737432f75b06ec734777a7/atmos_model.F90#L3510

uturuncoglu commented:
@DeniseWorthen Okay. Here are my steps for investigation and the possible solution,

- I edit `FV3/ccpp/driver/GFS_diagnostics.F90` file and add ofrac just after lfrac to check the fractions in FV3 side

```
    idx = idx + 1
    ExtDiag(idx)%axes = 2
    ExtDiag(idx)%name = 'ofrac'
    ExtDiag(idx)%desc = 'ocean fraction'
    ExtDiag(idx)%unit = 'fraction [0:1]'
    ExtDiag(idx)%mod_name = 'gfs_sfc'
    allocate (ExtDiag(idx)%data(nblks))
    do nb = 1,nblks
      ExtDiag(idx)%data(nb)%var2 => Sfcprop(nb)%oceanfrac(:)
    enddo
```

- and add following lines to diag_table to write them as fv3 output,

```
"gfs_sfc",     "lfrac",       "lfrac",     "fv3_history2d",  "all",  .false.,  "none",  2
"gfs_sfc",     "ofrac",       "ofrac",     "fv3_history2d",  "all",  .false.,  "none",  2
```

It seems that both are fine. So, FV3 side seems fine in terms of definition of ocean and land fractions.

**ofrac**
![ofrac](https://github.com/NOAA-EMC/noahmp/assets/1228999/d5bf10c8-c71f-4fca-acb7-710d07d4a79a)

**lfrac**
![lfrac](https://github.com/NOAA-EMC/noahmp/assets/1228999/7ce213b1-6a57-4c8e-bb0b-e151054a335d)

- Then, I decided to check `addLsmask2grid` routine since this is the place that sets the masking for the grid. I tried to set the mask directly to the oceanfrac as following,

```
lsmask(i,j) = int(GFS_data(nb)%SfcProp%oceanfrac(ix))
```

I know it is not correct but I just want to see what happens. It seems that the mask has no weird patterns in this case and looks fine in the mediator history file. So, I started thinking that maybe this logic 

https://github.com/NOAA-EMC/fv3atm/blob/062f7958ccd9bd97b8737432f75b06ec734777a7/atmos_model.F90#L3537 

has some issue. I think this is the central part of the grid definition and having bug in here seems not realistic since this is used in all the UFS configurations and probably someone could notice the issue before. So, I am little bit confused. 

- But this also lead to another observation since the interpolation with this change in `addLsmask2grid` was fine (land was 0 and ocean was 1, which is reverse the convention). Then I realized that the mask in lnd->atm is set to 1 by default for UFS since I think it assumes that the model will be coupled only with ocean, ice and wave. So, I add the second part in `med_map_routehandles_initfrom_field()` routine for lnd->atm coupling direction.

```
    if (trim(coupling_mode(1:3)) == 'ufs') then
       if (n1 == compatm .and. n2 == complnd) then
          srcMaskValue = ispval_mask
          dstMaskValue = ispval_mask
       end if
       if (n1 == complnd .and. n2 == compatm) then
          srcMaskValue = ispval_mask
          dstMaskValue = 0
       end if
    end if
```

I am not sure how I missed this in the first place since it seems it is a central part of the land implementation. Maybe, I focused to data atmosphere and side-by-side configurations too much and missed this one.

Anyway, my summary is that it seems that 

- we have issue in `addLsmask2grid` call with following logic. I am not sure about the root cause of it but maybe you have some explanation.

```
!$omp parallel do default(shared) private(i,j,nb,ix)
    do j=jsc,jec
      do i=isc,iec
        nb = Atm_block%blkno(i,j)
        ix = Atm_block%ixp(i,j)
! use land sea mask: land:1, ocean:0
        lsmask(i,j) = floor(one + epsln - GFS_data(nb)%SfcProp%oceanfrac(ix))
      enddo
    enddo
```

- I think we have also issue in CMEPS side in terms of definition of mask in lnd->atm direction, which can be fixed with the way that I mentioned previously.

DeniseWorthen commented:
@uturuncoglu I think at this point, it would be really helpful to have a call about these issues, but right now this is how I'm thinking it through.

Let's leave LND out for now and just deal w/ ATM-OCN-ICE (Ice here doesn't really matter, since it's mask == ocean mask). Also, mapping and merging are different things (I know you know this, but I'm trying to separate the two requirements in my own mind as I think through it). 

The lsmask value should be the element mask. It is the value which, when used w/ the src and dst masking, determines which values are mapped and which values are masked between components. 

So for mapping: 

- The ATM exists everywhere, so on the source side, it's srcMask = spval. The atm mask here doesn't really matter.

- On the destination side, I think the ATM mask needs to be 1 anywhere the mapped ocean mask>0. I'm not sure this would be the same in CESM. But in UFS, ATM is expecting SST over any partial grid cell, so that it can calculate the fluxes over the watery part. Those values are then merged to OCN (using the ocean fraction in the ATM, the Sa_aofrac field). 

I don't think the lsmask above gives 1 when mapped ocean mask>0? If oceanfrac==mapped ocean mask and you have some non-zero ocean fraction, then doesn't the floor(1-oceanfrac)=0?

barlage commented:
@DeniseWorthen to your earlier point, the land model doesn't care about the land fraction, the land model runs for any grid that has some land, then the composite code combines the fluxes, etc. based on the fraction

uturuncoglu commented:
@DeniseWorthen yes. That would be great. If you could setup a call we could discuss more. I am open after 11am MT. @barlage JFYI, the land fraction used internally by CMEPS for interpolation.

uturuncoglu commented:
@DeniseWorthen oceanfrac variable is 0 over land and 1 over ocean and between in coastal regions. Here are my test with the algorithm,

```
 oceanfrac =   0.000000000000000E+000 lsmask =            1
 oceanfrac =   8.000000000000000E-002 lsmask =            0
 oceanfrac =   0.100000000000000      lsmask =            0
 oceanfrac =   0.500000000000000      lsmask =            0
 oceanfrac =   0.700000000000000      lsmask =            0
 oceanfrac =    1.00000000000000      lsmask =            0
```

So, it seems that this algorithm only set land sea mask to 1 if the ocean fraction is 0, which seems correct to me. I put some print statement to call to check the values in our case.

DeniseWorthen commented:
@uturuncoglu Yes, I also checked and it does work.

uturuncoglu commented:
@DeniseWorthen I also add following at the end of the `addLsmask2grid` call

```
    call ESMF_GridWriteVTK(fcstGrid, filename="fv3_grid", rc=rc)
    if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
```

To check the mask and it seems it is not correct and has bunch of zeros over land. I was thinking that maybe data type issue in here but it looks good to me since mask variable in the grid is integer and consistent with the variables defined in here. I am not sure lsmask needs to be initialized or not. I'll try and see what happens.

DeniseWorthen commented:
@uturuncoglu I looked at the ofrac on the atm from the mediator history. This seem really odd!

<img width="400" alt="Screenshot 2024-03-12 at 1 52 17 PM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/95248fda-f2e8-4f2e-b668-9d9683da6d08">

uturuncoglu commented:
@DeniseWorthen Yes. I am running same case on Orion now to see any issue over there or not. Maybe this is specific to Derecho. I wonder if you test this in an other platform? In the development and validation of land component, I always used Orion.

uturuncoglu commented:
@DeniseWorthen I have good news. I change the code like following,

```
-        lsmask(i,j) = floor(one + epsln - GFS_data(nb)%SfcProp%oceanfrac(ix))
+        lsmask(i-isc+1,j-jsc+1) = floor(one + epsln - GFS_data(nb)%SfcProp%oceanfrac(ix))
```

and now I could get the land sea mask correctly some sort. I still have some issue but 

```
    isc = GFS_control%isc
    iec = GFS_control%isc+GFS_control%nx-1
    jsc = GFS_control%jsc
    jec = GFS_control%jsc+GFS_control%ny-1
    allocate(lsmask(isc:iec,jsc:jec))
```

maybe above part of code is returning differently when you setup `cplflx = true`. In my case, it seems that `lsmask` using local indexes but in the coupled app it uses global. I did not check the how `GFS_control` indexes being set yet since Derecho is acting weird and I think there is a disk issue. The following is the mask from FV3, there are some large numbers but maybe this is system issue since just after this run, it kicked me out.

![Screenshot 2024-03-12 at 2 38 43 PM](https://github.com/NOAA-EMC/noahmp/assets/1228999/66ae51b3-f4e5-4ab6-912a-4c0ff4b9ec91)

DeniseWorthen commented:
@uturuncoglu That's great progress, but very puzzling. It's hard to imagine a difference in indexing between coupled/uncoupled. Let me try your fix too. 

On my side, I was able to run the atmlnd case with V2 surface files. It ran, but when I tried it debug mode, I got another seg fault like before:

```
forrtl: severe (408): fort: (3): Subscript #1 of the array LAIM_TABLE has value 0 which is less than the lower bound of 1

Image              PC                Routine            Line        Source
fv3.exe            000000000E2ECA16  lnd_comp_types_mp         731  lnd_comp_types.F90
```

uturuncoglu commented:
@DeniseWorthen Derecho is still acting weird. So, I could not double checked. Based on your test, it seems that we have still inconsistencies. If you don't mind could you try with https://github.com/NOAA-EMC/noahmp/hotfix/debug_mode in noahmp side. If you still getting error. Do not worry, I could look at.

DeniseWorthen commented:
In my debug compile, there is still a problem if I add your change:

```
forrtl: severe (408): fort: (3): Subscript #2 of the array LSMASK has value 1 which is less than the lower bound of 61

Image              PC                Routine            Line        Source
fv3.exe            00000000022B6E42  atmos_model_mod_m        3542  atmos_model.F90
```

Do I need to allocate lsmask w/ different bounds?

uturuncoglu commented:
@DeniseWorthen Let me try in my side.

DeniseWorthen commented:
@uturuncoglu The hot fix branch runs in debug mode + v2 surface files. But, if I look at the mediator fractions, all I see are coastal points w/ non-zero values! This is w/o your lsmask index change. So the med_frac_lnd_lfrac now looks like

<img width="400" alt="Screenshot 2024-03-12 at 6 15 47 PM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/57e5af82-57ee-47ce-b498-dcf20f22cbab">

uturuncoglu commented:
@DeniseWorthen Okay. I am trying to run in my side too. Land frac looks weird but maybe it is related with the mask issue in the CMEPS side. You could try with following change in CMEPS med_map_routehandles_initfrom_field() call,

    if (trim(coupling_mode(1:3)) == 'ufs') then
       if (n1 == compatm .and. n2 == complnd) then
          srcMaskValue = ispval_mask
          dstMaskValue = ispval_mask
       end if
       if (n1 == complnd .and. n2 == compatm) then
          srcMaskValue = ispval_mask
          dstMaskValue = 0
       end if
    end if

So, this might solve the issue.

DeniseWorthen commented:
@uturuncoglu I had the same thought, and ran that case too. And now yes, the fractions look reasonable. And the ``atmexp_sl_zfun`` now matches what the land is sending

<img width="400" alt="Screenshot 2024-03-12 at 6 52 31 PM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/280b8832-f2cc-49a6-a8a7-f11dbbbab3f9">

uturuncoglu commented:
@DeniseWorthen I think with the mask changes in the CMEPS side, it is working without any issue with debug mode. It seems also issue with land fraction is gone. Maybe it is related with the debug mode. Is this compiles the code without optimization?

uturuncoglu commented:
@DeniseWorthen Same time. It is good to know that we have same results. The next thing is to run it without debug mode to see land fraction is fine or not. If we have issue with fraction and them maybe optimization is playing some role in here.

uturuncoglu commented:
@DeniseWorthen Land sea mask from the VTK files are also good with debug. I am compiling optimized one to see what happens.

DeniseWorthen commented:
So what really have we changed here? We're not changing the lsmask in atmos_model, but you've fixed a couple of bugs in noahmp and we've changed the mapping masks. But you were testing all of that earlier today, when you were getting bad VTK files?

Is the only difference now that we're using v2 surface files in addition to the other changes? 

I wanted to see if the fix is really coming from v2 surface files. I tried to run the same executable (hotfix + cmeps mapping) w/ the old surface files and I still get a set fault 

```
forrtl: severe (408): fort: (3): Subscript #1 of the array ALBSAT_TABLE has value -999 which is less than the lower bound of 1

Image              PC                Routine            Line        Source
fv3.exe            000000000AB0A142  noahmpdrv_mp_tran        1402  noahmpdrv.F90
```

uturuncoglu commented:
@DeniseWorthen The fix does not fix the issue with the old surface files since we are still getting soil color from fixed file not from sfc file. In v2 case, those files are consistent (soil color too) and it is working. I did not check that the v2 sfc files has soil color in it. If so, I could get that one too. Anyway, the code crashes since old files are not consistent in terms of soil color.

DeniseWorthen commented:
@uturuncoglu I don't think it necessary to make the current code work w/ the old input, I'm just wondering what really "fixed" the land mask issue?

We definitely need to commit a debug atm-land test.

uturuncoglu commented:
@DeniseWorthen Okay. Good news. It seems that optimized one also working fine. We have no land fraction issue anymore. Please double check in your end too if you could find time. Anyway, let me work on little bit more in noahmp fix branch and clean it and if soil color is available in sfc file I could get it from there too. I'll update you about it. In the mean time you could get the mask change to your CMEPS PR. I also need to run other cases to be sure that sbs and datm coupled ones are fine. I am not expecting any issue but just to be in the safe side. I think with this changes, we will have answer change in the land related tests. I am also plaining to add land debug test. Do you want me to crate ufs-weather-model branch that you could get it to your PR?

DeniseWorthen commented:
Good morning @uturuncoglu. I was hoping to bring in the cpl_scalars work to UFS in a single PR w/ no baseline changes (except for the single history file which is compared and which now is on tiles). So I'd like to do the following if it works for you.

1) update my ESCOMP CMEPS PR w/ the two changes (the mapping masks for LND and getting the fractions on tiles when you use the single mediator history phase). This probably means you'll need to re-test in CESM.

2) Commit a UFS PR containing the cpl_scalars for FV3, NoahMP and CMEPs. 

3) You can have more time to confirm all the NoahMP bug fixes and V2 surface files are working correcting. Then we create a UFS PR w/ the NoahMP bug fixes + debug test + v2 surface files. Once you give me the go-ahead, I'm happy to do that and  do all the testing to get it into the commit Q. 

Let me know if this all works for you.

uturuncoglu commented:
@DeniseWorthen yes, that works. Then, I could create fix PR. BTW, It seems everything is fine except 2m specific humidity. We have some small values around coastline. I am not sure this is v2 data files issue or not. Are you seeing same issue in your end?

DeniseWorthen commented:
Let me check. I didn't specifically look at q2m.

DeniseWorthen commented:
The 2m sp hum looks wrong in the sfc tile. In the mediator history, the atm import and the lnd export agree though.
<img width="800" alt="Screenshot 2024-03-13 at 9 58 03 AM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/e57c44ee-78ab-41e9-bf6c-43e0846e9cf1">

My test case is V2 + noahmp hotfix + cmeps mapping change.

uturuncoglu commented:
@DeniseWorthen Thanks for checking. Did you also look at t2m? That has no similar issue. Anyway, there might be a bug in CCPP Level too. So, I'll check it. Anyway, please ad any piece to your PR and I could start working from that point. Thanks for all your help.

DeniseWorthen commented:
I see the T2m temperature in the sfc file showing the same 0.0 values along the coast as the specific humidity. 

I didn't want to push the two fixes to my ESCOMP CMEPS PR in case there is still an issue in CMEPS mapping. I don't think there is, because the lnd import and atm exports look OK. But I wanted to check w/ you first.

uturuncoglu commented:
@DeniseWorthen Okay it is good to know we have issue in t2m. In my case, I could not see it. How did you do it? BTW, I think CMEPS mask change for lnd->atm is fine.

uturuncoglu commented:
@DeniseWorthen In my case, if I check sfc file the `tmpsfc` seems fine.

uturuncoglu commented:
@DeniseWorthen Okay. Got it. `tmp2m` has issue. The q2m and t2m follows different path in CCPP. So, there could be issue over there since I am not seeing any issue in latent and sensible heat.

uturuncoglu commented:
@DeniseWorthen BTW, i am plaining to open new issue in UFS Weather Model side to track the issue specific to lnd->atm coupling and also inconsistencies about the input files. I could use it also in the fix PR that will come soon.

DeniseWorthen commented:
@uturuncoglu I'm guessing at how to match mediator history variables and surface file variables. But this is T2m from sfc file and sa_ta from the mediator

<img width="400" alt="Screenshot 2024-03-13 at 11 57 51 AM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/2fd2cf35-c917-4a42-871b-606618aca709">

and tmpsfc and tskn from the mediator
<img width="400" alt="Screenshot 2024-03-13 at 11 57 25 AM" src="https://github.com/NOAA-EMC/noahmp/assets/40498404/3ee00063-f165-480f-9182-6b81f5885fba">

uturuncoglu reviewed (approved):
This looks fine. Thanks for adding `cpl_scalars` support. The GitHub action is also passing for DATM+LND configuration. @barlage please let me know if you want to review this PR.

uturuncoglu commented:
@DeniseWorthen @barlage JFYI, I create a new issue under UFS Weather Model to resolve the issues about inconsistent input and also debug mode run. https://github.com/ufs-community/ufs-weather-model/issues/2189

DeniseWorthen commented:
@uturuncoglu Great. I've run the ``control_p8_atmlnd`` with the CMEPS branch containing the mapping difference and it does change the baseline. Not unexpected, but just so you know.

uturuncoglu commented:
@DeniseWorthen Yes. Thanks for confirming. If I remember correctly, datm coupled configurations and sbs were fine with this change in my tests. Anyway, I am working on it.

uturuncoglu commented:
@DeniseWorthen I created a draft PR in here https://github.com/ufs-community/ufs-weather-model/pull/2191. This will go after yours since depends on CMEPS changes.

DeniseWorthen commented:
@uturuncoglu Thanks for your followup work on this. Let me know if I can help to test. I'll create the final sync-to-emc CMEPS PR once you merge (assuming no issues).

DeniseWorthen committed:
remove unused NUOPC use statement

zach1221 commented:
Hello, @barlage . Testing is complete on WM PR [#2175](https://github.com/ufs-community/ufs-weather-model/pull/2175). Could you please merge this noahmp PR for us?

DeniseWorthen commented:
Maybe @uturuncoglu can merge also?

uturuncoglu commented:
@DeniseWorthen @zach1221 Okay. I merge it.
